design/backup-resource-list.md

```diff
@@ -0,0 +1,92 @@
+# Expose list of backed up resources in backup details
+
+Status: Draft
+
+To increase the visibility of what a backup might contain, this document proposes storing metadata about backed up resources in object storage and adding a new section to the detailed backup description output to list them.
+
+## Goals
+
+- Include a list of backed up resources as metadata in the bucket
+- Enable users to get a view of what resources are included in a backup using the Velero CLI
+
+## Non Goals
+
+- Expose the full manifests of the backed up resources
+
+## Background
+
+As reported in #396, the information reported in a `velero backup describe <name> --details` command is fairly limited, and does not easily describe what resources a backup contains.
+In order to see what a backup might contain, a user would have to download the backup tarball and extract it.
+This makes it difficult to keep track of different backups in a cluster.
+
+## High-Level Design
+
+After performing a backup, a new file will be created that contains the list of the resources that have been included in the backup.
+This file will be persisted in object storage alongside the backup contents and existing metadata.
+
+A section will be added to the output of `velero backup describe <name> --details` command to view this metadata.
+
+## Detailed Design
+
+### Metadata file
+
+This metadata will be in JSON (or YAML) format so that it can be easily inspected from the bucket outside of Velero tooling, and will contain the API resource and group, namespaces and names of the resources:
+
+```
+deployments.apps:
+- default/database
+- default/wordpress
+services:
+- default/database
+- default/wordpress
+secrets:
+- default/database-root-password
+- default/database-user-password
+configmaps:
+- default/database
+```
+
+The filename for this metadata will be `<backup name>-resource-list.json`.
+The top-level key is the string form of the `schema.GroupResource` type that we currently keep track of in the backup controller code path.
+
+### Changes in Backup controller
+
+The Backupper currently initialises a map to track the `backedUpItems` (https://github.com/heptio/velero/blob/1594bdc8d0132f548e18ffcc1db8c4cd2b042726/pkg/backup/backup.go#L269), this is passed down through GroupBackupper, ResourceBackupper and ItemBackupper. Moving the map initialisation to the BackupController will provide access to the resulting list there after a successful backup.
```

Comment:
We're already a little bit inconsistent with how we return information from `pkg/backup` to the backup controller, but I think my preferred approach here would be to add a field to the `pkg/backup.Request` struct to store backed-up items. This can get populated by `pkg/backup` during execution of the `Backup(...)` method, and then the backup controller can read it when it goes to persist.  This is consistent with how volume snapshot metadata is passed currently.  

Alternately, we could define a new struct to capture returned info about the backup that would include volume snapshots, restic backup data (WIP by carlisia), and this info.

WDYT?

Reply:
Ah, I thought the backup.Request was the CR, but I see now that it's a different internal struct. I agree that it makes sense to use this or another struct to pass down the result.

Defining a new struct (i.e. backup.Response) will help make it more clear what it is used for, but using the Request struct isn't too terrible and won't require as many changes. What do others think?

Reply:
👍using the `backup.Request` WFM.